test/unit/extra_fields_test.rb

```diff
@@ -47,233 +42,213 @@ class ExtraFieldsTest < ActiveSupport::TestCase
-        assert @buyer.defined_builtin_fields.map(&:name).exclude?("extra_field")
-      end
-
-      should '#fields_to_xml not include extra fields' do
-        @buyer.send "#{@provider_field}=", "builtin"
-        @buyer.extra_fields = { "extra_field" => "extra_field" }
-        @buyer.save!
-        xml = Builder::XmlMarkup.new
-
-        @buyer.fields_to_xml(xml)
-        assert xml.to_s !~ /extra_field/
-      end
+    test 'FieldsMethods #defined_builtin_fields not include extra fields' do
+      FactoryBot.create(:fields_definition, account: @provider, target: 'Account', name: 'extra_field', required: true)
+
+      assert @buyer.defined_builtin_fields.map(&:name).exclude?('extra_field')
     end
 
-    context 'validations' do
+    test 'FieldsMethods #fields_to_xml not include extra fields' do
+      FactoryBot.create(:fields_definition, account: @provider, target: 'Account', name: 'extra_field', required: true)
 
-      context 'buyer resource' do
+      @buyer.send "#{@provider_field}=", 'builtin'
+      @buyer.extra_fields = { 'extra_field' => 'extra_field' }
+      @buyer.save!
+      xml = Builder::XmlMarkup.new
 
-        should 'not be done by default' do
-          assert @buyer.valid?
-          assert @buyer.errors[@provider_field].empty?
+      @buyer.fields_to_xml(xml)
+      assert xml.to_s !~ /extra_field/
+    end
+
+    test 'Source be provider fields for existing providers' do
+      @provider.send "#{@master_field}=", nil
+      @provider.validate_fields!
+      assert @provider.valid?
+
+    end
+
+    test 'Source be master fields for new providers' do
+      new_provider = FactoryBot.build(:provider_account)
+      assert new_provider.defined_fields.map(&:name).include?(@master_field)
+    end
+
+    test 'RequiredFields be created automatically for provider' do
+      provider = FactoryBot.create(:provider_account)
+
+      assert provider.fields_definitions.present?
+      assert provider.fields_definitions.all?(&:required?)
+    end
+
+    test 'Validations never be done for provider resource' do
+      @provider.send "#{@provider_field}=", nil
+      @provider.validate_fields!
+
+      assert @provider.field_value(@provider_field).nil?
+      assert @provider.valid?
+    end
+
+    test 'Validations never be done for master resource' do
+      master_account.send "#{@master_field}=", nil
+      master_account.validate_fields!
+
+      assert master_account.field_value(@master_field).nil?
+      assert master_account.valid?
+    end
+
+    test 'Validations BuyerResource resource not be done by default' do
+      assert @buyer.valid?
+      assert @buyer.errors[@provider_field].empty?
+    end
+
+    test 'Validations BuyerResource resource be done if said so' do
+      @buyer.validate_fields!
+
+      assert false == @buyer.valid?
+      assert @buyer.errors[@provider_field].present?
+    end
+
+  end
+
+  class Validations < FieldsDefinitionTest
```

Comment:
Do all these 3 classes need to inherit from `FieldsDefinitionTest`?

Reply:
@guicassolato Yeah, I do not like it either so I am refactoring it step by step in different commits :smile: